ee/tables/homebrew/upgradeable.go

```diff
@@ -0,0 +1,146 @@
+// +build !windows
+
+package brew_upgradeable
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"log/slog"
+	"os/exec"
+	"os/user"
+	"strconv"
+	"strings"
+	"syscall"
+
+	"github.com/kolide/launcher/ee/allowedcmd"
+	"github.com/kolide/launcher/ee/dataflatten"
+	"github.com/kolide/launcher/ee/tables/dataflattentable"
+	"github.com/kolide/launcher/ee/tables/tablehelpers"
+	"github.com/osquery/osquery-go/plugin/table"
+)
+
+const allowedCharacters = "0123456789"
+
+type Table struct {
+	slogger *slog.Logger
+	execCC  allowedcmd.AllowedCommand
+}
+
+func TablePlugin(slogger *slog.Logger) *table.Plugin {
+	columns := dataflattentable.Columns(
+		table.TextColumn("uid"),
+	)
+
+	t := &Table{
+		slogger: slogger.With("table", "kolide_brew_upgradeable"),
+		execCC:  allowedcmd.Brew,
+	}
+
+	return table.NewPlugin("kolide_brew_upgradeable", columns, t.generate)
+}
+
+func (t *Table) generate(ctx context.Context, queryContext table.QueryContext) ([]map[string]string, error) {
+	var results []map[string]string
+
+	uids := tablehelpers.GetConstraints(queryContext, "uid", tablehelpers.WithAllowedCharacters(allowedCharacters))
+	if len(uids) < 1 {
+		return results, fmt.Errorf("kolide_brew_upgradeable requires at least one user id to be specified")
+	}
+
+	for _, uid := range uids {
+		for _, dataQuery := range tablehelpers.GetConstraints(queryContext, "query", tablehelpers.WithDefaults("*")) {
+			output, err := t.getBrewOutdated(ctx, uid)
+			if err != nil {
+				t.slogger.Log(ctx, slog.LevelInfo, "failure querying user brew installed packages", "err", err, "target_uid", uid)
+				continue
+			}
+
+			flattenOpts := []dataflatten.FlattenOpts{
+				dataflatten.WithSlogger(t.slogger),
+				dataflatten.WithQuery(strings.Split(dataQuery, "/")),
+			}
+
+			flattened, err := dataflatten.Json(output, flattenOpts...)
+			if err != nil {
+				t.slogger.Log(ctx, slog.LevelInfo, "failure flattening output", "err", err)
+				continue
+			}
+
+			rowData := map[string]string{
+				"uid": uid,
+			}
+
+			results = append(results, dataflattentable.ToMap(flattened, dataQuery, rowData)...)
+		}
+	}
+
+	return results, nil
+}
+
+func (t *Table) getBrewOutdated(ctx context.Context, uid string) ([]byte, error) {
+	cmd, err := t.execCC(ctx, "outdated", "--json")
+	if err != nil {
+		return nil, fmt.Errorf("creating brew outdated command: %w", err)
+	}
+
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return nil, fmt.Errorf("assigning StdoutPipe for brew outdated command: %w", err)
+	}
+
+	if err := runAsUser(ctx, uid, cmd); err != nil {
+		return nil, fmt.Errorf("runAsUser brew outdated command as user %s: %w", uid, err)
+	}
+
+	data, err := io.ReadAll(stdout)
+	if err != nil {
+		return nil, fmt.Errorf("ReadAll brew outdated stdout: %w", err)
+	}
+
+	if err := cmd.Wait(); err != nil {
+		return nil, fmt.Errorf("deallocation of brew outdated command as user %s: %w", uid, err)
+	}
+
+	return data, nil
+}
+
+func runAsUser(ctx context.Context, uid string, cmd *exec.Cmd) error {
```

Comment:
We use this in enough places now that we should probably just make this available/reusable somewhere. I don't think it's necessary to address in this PR, but wanted to flag in case you (or James/Zack/seph) have a good suggestion for where we should move this in the future.

Reply:
Agreed. I feel fried this week, so I too didn't want to address it in this PR, but I might take a look at that next week if I can.

Reply:
Agree it feels like we should put it into a helper. We have some exec helpers. Which reminds me, this isn't using them. So whether we expand the exec helper, or make another 🤷